imports/plugins/core/ui-tagnav/client/containers/tagNavContainer.js

```diff
@@ -0,0 +1,389 @@
+  Visible: "stateNavbarVisible"
+};
+
+const NavbarOrientation = {
+  Vertical: "vertical",
+  Horizontal: "horizontal"
+};
+
+const NavbarVisibility = {
+  Shown: "shown",
+  Hidden: "hidden"
+};
+
+const NavbarPosition = {
+  Static: "static",
+  Fixed: "fixed"
+};
+
+const NavbarAnchor = {
+  Top: "top",
+  Right: "right",
+  Bottom: "bottom",
+  Left: "left",
+  None: "inline"
+};
+
+const TagNavHelpers = {
+  onTagCreate(tagName, parentTag) {
+    TagHelpers.createTag(tagName, undefined, parentTag);
+  },
+  onTagRemove(tag, parentTag) {
+    TagHelpers.removeTag(tag, parentTag);
+  },
+  onTagSort(tagIds, parentTag) {
+    TagHelpers.sortTags(tagIds, parentTag);
+  },
+  onTagDragAdd(movedTagId, toListId, toIndex, ofList) {
+    TagHelpers.moveTagToNewParent(movedTagId, toListId, toIndex, ofList);
+  },
+  onUpdateTag(tagId, tagName, parentTagId) {
+    TagHelpers.updateTag(tagId, tagName, parentTagId);
+  },
+  isMobile() {
+    return window.matchMedia("(max-width: 991px)").matches;
+  },
+  tagById(tagId, tags) {
+    return _.find(tags, (tag) => tag._id === tagId);
+  },
+  updateSuggestions(suggestion, excludeTagsObj) {
+    return TagHelpers.updateSuggestions(suggestion, excludeTagsObj);
+  },
+  hasSubTags(tagId, tags) {
+    const foundTag = this.tagById(tagId, tags);
+
+    if (foundTag) {
+      if (Array.isArray(foundTag.relatedTagIds) && foundTag.relatedTagIds.length) {
+        return true;
+      }
+    }
+    return false;
+  }
+};
+
+let editable = false;
+let selectedTag = null;
+
+class TagNavContainer extends Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      attachedBodyListener: false,
+      editable: editable,
+      tagIds: props.tagIds || [],
+      tagsByKey: props.tagsByKey || {},
+      selectedTag: selectedTag,
+      suggestions: [],
+      [NavbarStates.Visible]: false,
+      [NavbarStates.Visibile]: NavbarVisibility.Hidden,
+      newTag: {
+        name: ""
+      }
+    };
+  }
+
+  componentDidMount() {
+    $(window).on("resize", this.onWindowResize).trigger("resize");
+  }
+
+  componentWillReceiveProps(nextProps) {
+    const { tagIds, tagsByKey } = nextProps;
+    this.setState({ tagIds, tagsByKey });
+  }
+
+  componentWillUnmount() {
+    $(window).off("resize", this.onWindowResize);
+  }
+
+  onWindowResize = () => {
+    if (window.matchMedia("(max-width: 991px)").matches) {
+      this.setState({
+        [NavbarStates.Orientation]: NavbarOrientation.Vertical,
+        [NavbarStates.Position]: NavbarPosition.Fixed,
+        [NavbarStates.Anchor]: NavbarAnchor.Left
+      });
+    } else {
+      this.setState({
+        [NavbarStates.Orientation]: NavbarOrientation.Horizontal,
+        [NavbarStates.Position]: NavbarPosition.Static,
+        [NavbarStates.Anchor]: NavbarAnchor.None,
+        [NavbarStates.Visible]: false
+      });
+    }
+  }
+
+  handleNewTagSave = (tag, parentTag) => {
+    editable = true; // keep editing state after re-render
```

Comment:
You shouldn't edit `state` directly.
https://facebook.github.io/react/docs/state-and-lifecycle.html#do-not-modify-state-directly

`this.setState()` causes a re-render every time it is encountered.